core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,396 @@
+    }
+    if (cmk_blob == NULL || SIZE_OF_KEYBLOB_T != real_blob_len)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint8_t *tmp = (uint8_t *)malloc(keysize);
+    if (tmp == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(tmp, keysize);
+    if (ret != SGX_SUCCESS)
+    {
+        free(tmp);
+        return ret;
+    }
+    ret = ehsm_create_keyblob(tmp,
+                              keysize,
+                              NULL,
+                              0,
+                              (sgx_aes_gcm_data_ex_t *)cmk_blob);
+
+    memset_s(tmp, keysize, 0, keysize);
+
+    free(tmp);
+    return ret;
+}
+
+sgx_status_t ehsm_create_asymmetric_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_SUCCESS;
+
+    EVP_PKEY_CTX        *pkey_ctx        = NULL;
+    EVP_PKEY            *pkey           = NULL;
+    EC_GROUP            *ec_group       = NULL;
+    BIO                 *bio            = NULL;
+    uint8_t             *pem_keypair    = NULL;
+    uint32_t            key_len         = 0;
+
+    switch (cmk->metadata.keyspec) {
+        case EH_RSA_2048:
+        case EH_RSA_3072:
+        case EH_RSA_4096:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+            break;
+        case EH_EC_P224:
+        case EH_EC_P256:
+        case EH_EC_P384:
+        case EH_EC_P512:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+            break;
+        case EH_SM2:
+            pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_SM2, NULL);
+            break;
+        default:
+            break;
+    }
+    if (!pkey_ctx) {
+        goto out;
+    }
+
+    if (!EVP_PKEY_keygen_init(pkey_ctx)) {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec) {
+        case EH_RSA_2048:
+            EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_ctx, RSA_2048_KEY_BITS);
+            break;
+        case EH_RSA_3072:
+            EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_ctx, RSA_3072_KEY_BITS);
+            break;
+        case EH_RSA_4096:
+            EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_ctx, RSA_4096_KEY_BITS);
+            break;
+        case EH_EC_P224:
+        case EH_EC_P256:
+        case EH_EC_P384:
+        case EH_EC_P512:
+            EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_X9_62_prime256v1);
+            break;
+        case EH_SM2:
+            EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_sm2);
+            break;
+        default:
+            break;
+    }
+    
+    if (!EVP_PKEY_keygen(pkey_ctx, &pkey)) {
+        goto out;
+    }
+
+    EVP_PKEY_CTX_free(pkey_ctx);
+
+    if (!(bio = BIO_new(BIO_s_mem()))) {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PUBKEY(bio, pkey)) {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PrivateKey(bio, pkey, NULL, NULL, 0, NULL, NULL)) {
+        goto out;
+    }
+
+    key_len = BIO_pending(bio);
+    if (key_len == 0) {
+        goto out;
+    }
+    if (cmk->keybloblen == 0) {
```

Comment:
the original purpose of the function ehsm_calc_keyblob_len is to handle the different case handling.

Reply:
it has been change